### PR TITLE
Fixes issue #1128, adds descs & makes controllers more flexible

### PR DESCRIPTION
**Pull request by Dahlgaard**
Fixes viewscreens #1128 and makes controllers more flexible

Added description to viewscreen & viewscreen_controller

Added base call to the Initialize() of viewscreen. 
    This is required for objects or they must set initialized to true themselves. 
    Failing to do this means you cannot click it as it will be rejected in click.dm. 
    This was causing problems referenced in issue #1128

Removed explicit Initialize() for viewscreen_controller. 
    This code was never set up to handle more then 1 viewscreen in the world, the current ship has 2. 
    The result was that all viewscreen_controllers were linked to the viewscreen on the backup battle bridge. 
    Possibly also part of items mentioned in issue #1128

Added helper proc get_area_viewscreens() which returns a list of viewscreens in the area.

Changed viewscreen_controller/attack_hand. 
    This will now popup a screen with a list of all viewscreens in the current area, allowing users to control them all from one viewscreen controller panel. 
    We can now have more viewscreens on the bridge if we want.

Changed viewscreen_controller/Topic. 
    Supports new actions in the ui. 
    Now has a usr.canUseTopic() check to ensure that non TK users cannot use the console from halfway across the ship.

### CODE REVIEW

**Comment by Dahlgaard** [comment]
note sure if the CL part is necessary of filled out correctly feel free to wipe / edit it.

**Comment by Dahlgaard** [comment]
CL part stripped after talks with ninja

File: code/game/objects/structures/viewscreen.dm [hunk 1]
```diff
@@ -124,25 +126,28 @@
...
+	for(var/obj/structure/viewscreen/V in area_contents(A))
+		screens += V
+	return screens
 
 /obj/structure/viewscreen_controller/attack_hand(mob/user)
 	var/dat = "<B>Viewscreen Control Panel</B><HR>"
-	dat += "Viewing Mode: [linked.view_mode ? "Navigation" : "Tactical"]"
-	dat += "<BR><BR><A href=?src=\ref[src];switch=1>Switch View Modes</A>"
+
+	for(var/obj/structure/viewscreen/V in get_area_viewscreens())
```
Comment (ninjanomnom) [inline code/game/objects/structures/viewscreen.dm]: You should not be re checking the area for viewscreen every time someone clicks the controller. Make a list on the controller holding all known screens instead and put a refresh button in the ui.

File: code/game/objects/structures/viewscreen.dm [hunk 2]
```diff
@@ -124,25 +133,35 @@
 
 /obj/structure/viewscreen_controller
 	name = "viewscreen control panel"
+	desc = "This is a control panel for the viewscreens"
 	icon = 'icons/obj/airlock_machines.dmi'
 	icon_state = "airlock_control_standby"
 
 	density = 0
 	anchored = 1
+	var/list/linked_viewscreens = null
```
Comment (ninjanomnom) [inline code/game/objects/structures/viewscreen.dm]: no need for `= null` here

File: code/game/objects/structures/viewscreen.dm [hunk 3]
```diff
@@ -124,25 +133,35 @@
...
 
-/obj/structure/viewscreen_controller/Initialize()
-	for(var/obj/structure/viewscreen/V in world)
-		linked = V
+/obj/structure/viewscreen_controller/attack_hand(mob/user)
 
-	..()
+	// If this is being called for the first time, populate the list of screens
+	if (!linked_viewscreens)
+		linked_viewscreens = get_area_viewscreens()
```
Comment (ninjanomnom) [inline code/game/objects/structures/viewscreen.dm]: do this in initialize

File: code/game/objects/structures/viewscreen.dm [hunk 4]
```diff
@@ -151,8 +170,16 @@
 
 /obj/structure/viewscreen_controller/Topic(href,href_list)
 	..()
-	if(href_list["switch"])
-		linked.view_mode = !linked.view_mode
+	if(!usr.canUseTopic(src))
```
Comment (ninjanomnom) [inline code/game/objects/structures/viewscreen.dm]: don't use `usr`

File: code/game/objects/structures/viewscreen.dm [hunk 5]
```diff
@@ -151,8 +170,16 @@
...
+		return
+
+	if(href_list["refreshscreens"])
+		linked_viewscreens = get_area_viewscreens()
+
+	if(href_list["switchmode"])
+		var/obj/structure/viewscreen/V = locate(href_list["switchmode"])
```
Comment (ninjanomnom) [inline code/game/objects/structures/viewscreen.dm]: instead of re locating the screen you should be passing the index of the screen in linked_viewscreens

**Comment by ninjanomnom** [comment]
Let me know if/when you work on this again

**Pull request closed without merging**